App V2: Update Look and Feel of Starred Builds to use Toggle Switch

biyeun opened the pull request:
@orangejenny cc @gcapalbo

fyi @proteusvacuum

<img width="1127" alt="screen shot 2017-06-05 at 4 16 38 pm" src="https://cloud.githubusercontent.com/assets/716573/26801280/61a36918-4a0a-11e7-982d-824c14d348ae.png">

<img width="344" alt="screen shot 2017-06-05 at 4 21 19 pm" src="https://cloud.githubusercontent.com/assets/716573/26801514/1fb5ca2c-4a0b-11e7-9ac0-c7618c757b6f.png">

biyeun committed:
use bootstrap-switch in app manager

biyeun committed:
enable bootstrap switch after template render

biyeun committed:
update starred builds / auto update to use toggle

biyeun committed:
remove defunc less

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/releases.js:
```diff
@@ -334,6 +334,17 @@ hqDefine('app_manager/js/releases.js', function () {
+            $(elem).find('.js-auto-update-checkbox').bootstrapSwitch({
+                state: savedApp.is_released(),
+                onSwitchChange: function () {
+                    self.toggleRelease(savedApp)
```
Missing semicolon. (semi)

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/releases.js:
```diff
@@ -334,6 +334,17 @@ hqDefine('app_manager/js/releases.js', function () {
+                state: savedApp.is_released(),
+                onSwitchChange: function () {
+                    self.toggleRelease(savedApp)
+                }
```
Missing trailing comma. (comma-dangle)

biyeun committed:
lint

orangejenny commented:
Hey, I'm a bit confused on this. I thought the feedback on the previous iteration of this was just that the legend in the top right looked clickable, not that the general interaction of clicking on a star, as people do in v1, is confusing. The idea, and the language, of "starring" a build is a rather convenient shorthand, and replacing it with "auto update" language seems clunky.

proteusvacuum commented:
Hey,
I somewhat agree with @orangejenny , however I do think this is an overall improvement over a "star" which has no meaning really. 

I have a few comments:
- I think in general with those switches it is much nicer to also show what will happen when you click it, so have `On|Off`, where off is greyed out, so that you are clicking on the action that you want to have happen (e.g. https://ghinda.net/css-toggle-switch/bootstrap.html)
- That top bar is very busy, and adding more words to it seems like a shame. I also think being able to scan down the side to see which are starred and aren't is nice to be able to do. What if you changed "On|Off" to something like "Released | Test Build" or something (wording to be worked out). This way we don't need a legend (I think if we need a legend, the UI element we've chosen isn't clear enough).

biyeun committed:
update released versions language

biyeun committed:
update release / auto update toggle

biyeun committed:
update bootstrap switch styling

biyeun commented:
@orangejenny @proteusvacuum updated this to Farid's recommendation. The starred build is a lot more ambiguous and requires way more documentation than a released / test switch imo. This feels more straightforward.

orangejenny commented:
Code looks good.

On the interaction: I agree with Farid that it's nicer to be able to see what both options are. The bootstrap switch is sleek, but the basic `.btn-group` applied to radio buttons would be more usable.

On release: can you get Clark/Kevin's opinion on whether we should time documentation updates to happen at the same time that this is released? (Or if they're fine with the docs continuing to reference stars/starring for now? I don't have a good sense of how often that shows up in docs.)

proteusvacuum commented:
Agreed with @orangejenny re the button group (or just showing what the other option is) also, can we make "released" be green, or some other positive looking colour? (the ⭐️ is nice because it is bright and yellow)

Like the updates so far!

snopoke commented:
nice, I like the outcome here. One note on the 'released' help text - it should also mention something like `or if the user clicks 'Update App'`

proteusvacuum commented:
I was thinking about this while brushing my teeth this morning. What if we generate the bitly app code when a build is switched to "Released" and show it on this page? I think that would be a nice lower-friction way of getting the app on your phone. (Or even having a mini QR code that you can click either scan or click to enlarge). 

Also, if this is going to be the "Publish" page, maybe the action should be "Published" instead of "Released".

biyeun commented:
Hi all, got back from vacation yesterday and still catching up with some stuff and performance evals, but I'm really loving the suggestions here. Thank you! Will be playing around with some of the ideas presented and will hopefully have something to look at shortly. @proteusvacuum @orangejenny @snopoke

biyeun commented:
@proteusvacuum I love that you were thinking of the deploy page while brushing your teeth. :joy: :heart:

biyeun committed:
use btn group instead of slider switch

biyeun committed:
Revert "use bootstrap-switch in app manager"

This reverts commit 50f359fd4e93feada99e4a6a58e70bde610356ef.

biyeun committed:
hide comment if no comment is filled in

biyeun commented:
@proteusvacuum there was some chatter about the release panels already being too high. Adding the barcode / bitly code might just be too much.

@orangejenny  I did the thing with hiding the comments unless you fill it in, removing one line of vertical space.
<img width="1126" alt="screen shot 2017-06-05 at 4 12 55 pm" src="https://cloud.githubusercontent.com/assets/716573/26801194/116fb410-4a0a-11e7-8c44-3839ba5b9795.png">

biyeun commented:
@orangejenny @proteusvacuum @snopoke see previous comment for the updated button group version of Released/In Test

biyeun committed:
update help copy

biyeun commented:
@snopoke here is the new help copy 
<img width="344" alt="screen shot 2017-06-05 at 4 21 19 pm" src="https://cloud.githubusercontent.com/assets/716573/26801497/11c51800-4a0b-11e7-94b1-9c53d8e1a05c.png">

orangejenny commented:
ooh, I like that in test vs released are in different colors (it's the little details...), and the more compressed spacing when there aren't comments

orangejenny reviewed (approved):
Code looks good. Still think it's worth checking in with @dimagi/product about documentation changes.

biyeun commented:
added the hold off on merge as we're going to bundle this with an additional feature to auto detect when a new build is possible and highlight the app name as a clickable region

biyeun committed:
highlight app name when changes are available to publish

biyeun committed:
remove publish button

biyeun committed:
add edit form button on form settings page

biyeun commented:
@orangejenny 

updated
- removed publish button
- added indication to publish when changes are available
- added edit form button on form settings pages

<img width="273" alt="screen shot 2017-06-13 at 12 58 12 am" src="https://user-images.githubusercontent.com/716573/27066730-3b532fe4-4fd4-11e7-8363-f3c68eafc68a.png">
<img width="284" alt="screen shot 2017-06-13 at 12 46 36 am" src="https://user-images.githubusercontent.com/716573/27066731-3b55faa8-4fd4-11e7-9405-0eb71edb58b0.png">

biyeun committed:
define global hqLayout

proteusvacuum commented:
yay! Edit Form!

I'm wondering if it will be hard to find how you download the app you've already published to your phone if that little banner isn't there?

orangejenny commented:
Nice! I really like the banner, I'd had trouble imagining what a good indicator would look like.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/apps_base.js:
```diff
@@ -19,4 +19,16 @@ $(function () {
     }
+
+    $(document).ajaxComplete(function(e, xhr, options) {
+        if (/edit_form_attr/.test(options.url) ||
```
The url matching seems potentially fragile as we add new urls (or switch existing urls to use ajax). What would you think of doing something like firing an additional request to check the version number? It'd be another request, but a short one, and would be more future-proof.

biyeun replied:
Yes, this sounds like a very good idea. I was following the lead from app preview, mainly.

benrudolph replied:
👍 sorry i definitely wrote the first iteration of this because it was the easiest thing to get out the door. the one issue i see with the additional request is that there will likely be a lot of false positives if the request is made here since this triggers after any ajax request is made. essentially we want this to update when someone hits the save button, but it felt really hacky to splice that in there

orangejenny replied:
not quite following - what's the case where false positives would happen? i was thinking of the version-checking because it's what we do on the releases page to verify that you can/can't make a new build.

benrudolph replied:
any other ajax call on the page would trigger this callback, so we'd check the version regardless if it were a save or not. (may naively trigger an infinite loop, the call to check the version will trigger this callback method again)

orangejenny replied:
Oh, I see. Seems like a question then of whether to store the set of "saving" urls or the set of "non-saving" urls (at least the version check url).

orangejenny commented:
What would you think of having the new changes indicator also show on page load if there have been changes since the last build?

biyeun commented:
@orangejenny I like the idea of having the indicator show up on page load. I'll work on that in a few. Thanks!

biyeun committed:
show publish banner on reload or if there are changes

biyeun commented:
this now detects publish changes on page load (and every 20 seconds) @orangejenny

biyeun committed:
lint

biyeun commented:
@orangejenny fyi, the publish button was in fact removed from formdesigner here

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -141,6 +150,34 @@ hqDefine('app_manager/js/app_manager.js', function () {
+        };
+        _checkPublishStatus();
+        // check publish status every 20 seconds
+        setInterval(_checkPublishStatus, 20000);
```
This could probably be less frequent given that the code is also checking for known app-modifying ajax calls below.

orangejenny commented:
@biyeun Nice. Code looks fine, looks like v2 diffs are out of date.

biyeun commented:
@orangejenny do you think this is g2g? I don't want to keep piling on the features here... :P

orangejenny commented:
It's good from my perspective. I think @krsdimagi wanted to communicate with the field team specifically about this one, would check that he's okay with merging.

biyeun committed:
remove preview app button in favor of preview toggle bar

biyeun commented:
@orangejenny moved the app preview button to a turquiose bar along the left side of app preview 
<img width="1149" alt="screen shot 2017-06-19 at 1 36 24 pm" src="https://user-images.githubusercontent.com/716573/27298338-380a1e98-54f6-11e7-97f3-8c71c65f2516.png">
<img width="909" alt="screen shot 2017-06-19 at 1 36 30 pm" src="https://user-images.githubusercontent.com/716573/27298340-380dc5b6-54f6-11e7-97c2-6d57dbed0926.png">
<img width="1155" alt="screen shot 2017-06-19 at 1 39 44 pm" src="https://user-images.githubusercontent.com/716573/27298339-380c9dc6-54f6-11e7-928c-1f69b4265c1e.png">
<img width="1" alt="screen shot 2017-06-19 at 1 39 47 pm" src="https://user-images.githubusercontent.com/716573/27298337-3807dade-54f6-11e7-941e-56551db53669.png">
<img width="49" alt="screen shot 2017-06-19 at 1 39 50 pm" src="https://user-images.githubusercontent.com/716573/27298341-380e22fe-54f6-11e7-841c-c46b8a6dc2fd.png">

biyeun committed:
remove X over phone in app preview

biyeun committed:
update app preview appearance in formbuilder full screen mode

biyeun commented:
@proteusvacuum 
took care of full screen formbuilder and preview

<img width="908" alt="screen shot 2017-06-19 at 2 29 49 pm" src="https://user-images.githubusercontent.com/716573/27299953-09b3702a-54fc-11e7-91c6-1b265ef950d2.png">

biyeun committed:
add UI notification

biyeun committed:
update v2 diffs

biyeun commented:
Should be set to go @orangejenny 
cc @proteusvacuum 

This is the notification
<img width="1091" alt="screen shot 2017-07-14 at 1 25 26 pm" src="https://user-images.githubusercontent.com/716573/28223196-75817b6c-6898-11e7-8b6d-d268c810078b.png">

biyeun committed:
notification should show regardless of build profile access

biyeun committed:
update diffs

biyeun commented:
merging since all prior changes were approved...

orangejenny commented:
looks great!
